independent-projects/tools/codestarts/codestarts.adoc

```diff
@@ -0,0 +1,220 @@
+
+*NOTE* The `codestart.yml` file and the directory structure follow the same principle, it can optionally contain a base and/or some language overrides.
+
+=== Project generation
+
+When generating a Quarkus project:
+
+. Codestarts to use are resolved depending on the input
+. The Codestarts' shared data is processed to make it available for all codestarts
+. The relevant files are processed:
+** Only the files that are related to the selected language are processed
+** The files are processed differently based on a naming convention.
+** The data is used to render Qute templates
+** We always process in this order: language, project, buildtool, config, custom, other.
+
+The data used to generate a specific codestart is a merge of:
+
+. The data of the codestart to generate
+. All codestarts "shared" data
+. The user input
+. Some specific post processing (e.g. adding dependencies)
+
+*NOTE* The data (shared or not) can also be specific to a language.
+
+=== Directory Structure
+
+*NOTE* `codestart.yml` is the only required file.
+
+* `codestart.yml` must be at the root of the codestart
+* `./base` contains all the files that will be processed
+* `./[languageName]` contains all the files that will be processed if the specified language has been selected (overriding base)
+
+=== codestart.yml
+
+codestart.yml:
+[source,yaml]
+----
+name: the unique name - REQUIRED
+ref: the reference name (not unique) to use for extension matching (else the name is used)
+type: the type [example (default), project, buildtool, language, config] or any custom type
+fallback: flag to indicate that it should be selected as fallback when no codestart has been selected with type - Only for base types
+preselected: flag to indicate that it should be pre-selected - Only for extra types.
+spec:
+  [base or language name]: the specification (base, java, ...)
+    data: a map of data to use only for this codestart
+    shared-data: a map of data to use accross all codestarts
+    dependencies: a list of dependencies to add when this is selected
+    test-dependencies: a list of test dependencies to add when this is selected
+----
+
+=== Files Naming Convention
+
+* containing `.tpl.qute` will be processed with Qute (`.tpl.qute` will be removed from the output file name).
+* containing `.include.qute` is used as a Qute template for inclusion (only available locally to the codestart).
+
+  Example: When using `{#include [name]}` in a template, it will look for `[name].include.qute` in the language dir, then in the base dir or the codestart.
+
+* an output strategy can be defined for specific files, it defines some specific logic on how to deal with those files.
+
+ Example: in the Quarkus project `codestart.yml`:
+[source,yaml]
+----
+output-strategy:
+  "pom.xml": smart-pom-merge # smartly merge pom.xml contents together
+  "README.md": append # they are appended together
+  "readme.md": forbidden # it will fail if found in any codestart
+  "README.adoc": forbidden
```

Comment:
forbidden..why ?

Reply:
I don't think we can have multiple `README` types merged together, we need to choose one or the other at project level, no?

Reply:
I guess :)